feat(aws-backup): Add arn attribute and grant method to backup vault

DaWyz opened the pull request:
`BackupVault` now have the `backupVaultArn` attribute available. It is also possible to import an existing `BackupVault` by `arn` using the `fromBackupVaultArn` method.

You can also grant permission to an existing grantee using the `grant` method.

See the example below giving an IAM Role permission to start a backup job:

```ts
const importedVault = backup.BackupVault.fromBackupVaultName(this, 'Vault', 'myVaultName');

const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com') });

importedVault.grant(role, 'backup:StartBackupJob');
```

Closes #14996. 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e104e74c1af619b4e29ba11e0d17ff9d761dfa34
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=08976188-ef0a-4371-acec-eec88f6b2f60%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

BenChaimberg commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -15,6 +15,19 @@ export interface IBackupVault extends IResource {
+   * Grant the actions defined in actions to the given grantee
+   * on this Backup Vault resource.
+   */
+  grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
```
are there any common permissions patterns that you've used with backup vaults that would warrant a more specific grant method?

DaWyz replied:
So far, it was just trigger a backup job. it's not anything complex (single action) so I would leave it as is for now.

BenChaimberg commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -108,27 +121,71 @@ export enum BackupVaultEvents {
+   * @param grantee Principal to grant right to
+   * @param actions The actions to grant
+   */
+  public grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant {
```
according to the [AWS docs](https://docs.aws.amazon.com/aws-backup/latest/devguide/creating-a-vault-access-policy.html), wildcard actions (`"*"`) are not allowed; please add a validation

BenChaimberg commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -85,3 +85,26 @@ const vault = new BackupVault(stack, 'Vault', {
+## Importing existing backup vault
+
+To import an existing backup vault into your CDK application, use the `BackupVault.fromBackupVaultArn` or `BackupVault.fromBackupVaultName` 
+factory method. Here is an example of giving a lambda role permission to start a backup job:
```
```suggestion
static method. Here is an example of giving a Lambda function permission to start a backup job:
```

BenChaimberg commented on packages/@aws-cdk/aws-backup/test/vault.test.ts:
```diff
@@ -135,6 +136,89 @@ test('defaults to all notifications', () => {
+  // GIVEN
+  let vaultName = 'myVaultName';
+  let vault = BackupVault.fromBackupVaultName(stack, 'Vault', vaultName);
+  let fn = new lambda.Function(stack, 'function', {
```
I'd prefer using an IAM role here instead of using the relatively complex Lambda grantable

BenChaimberg commented on packages/@aws-cdk/aws-backup/README.md:
````diff
@@ -85,3 +85,26 @@ const vault = new BackupVault(stack, 'Vault', {
+
+```ts
+import * as lambda from '@aws-cdk/aws-lambda';
+import { Stack } from '@aws-cdk/core';
+
+let stack = new Stack();
````
I appreciate the complete example here, but you can just use `this` as the scope. If you'd like to ensure compilation, please create a [default fixture](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#rosetta) that sets up a stack

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 67300e9fcddf1f501fa14d1063fc3bb9ed0a9e26
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=60ce3add-d988-4240-9605-7fe63ae1bdbd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 33cf463f9b6dec5cbabf321d4ed877f7f2cffa02
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b8e24594-aecc-407b-9b28-7de27c9ec078%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

BenChaimberg commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -90,21 +88,11 @@ that contains recovery points will fail.
+const importedVault = backup.BackupVault.fromBackupVaultName(this, 'Vault', 'myVaultName');
 
-vault.grant(fn, 'backup:StartBackupJob');
+importedVault.grant(fn, 'backup:StartBackupJob');
```
```suggestion
const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com') });
importedVault.grant(role, 'backup:StartBackupJob');
```

BenChaimberg commented on packages/@aws-cdk/aws-backup/rosetta/default.ts-fixture:
```diff
@@ -0,0 +1,26 @@
+import { Duration, RemovalPolicy, Stack } from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import * as backup from '@aws-cdk/aws-backup';
+import * as dynamodb from '@aws-cdk/aws-dynamodb';
+import * as events from '@aws-cdk/aws-events';
+import * as kms from '@aws-cdk/aws-kms';
+import * as lambda from '@aws-cdk/aws-lambda';
+import * as sns from '@aws-cdk/aws-sns';
+
+declare const fn: lambda.Function;
+declare const myKey: kms.Key;
+declare const myCoolConstruct: Construct;
+declare const myTable: dynamodb.Table;
+declare const myTopic: sns.Topic;
+declare const myVault: backup.BackupVault;
+declare const otherVault: backup.BackupVault;
```
are any of these imports/declarations used? please delete if not

BenChaimberg commented on packages/@aws-cdk/aws-backup/rosetta/default.ts-fixture:
```diff
@@ -0,0 +1,26 @@
+declare const myTopic: sns.Topic;
+declare const myVault: backup.BackupVault;
+declare const otherVault: backup.BackupVault;
+declare const plan: backup.BackupPlan;
```
create a separate fixture that provides a plan so that the default is not polluted

BenChaimberg commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -59,29 +57,42 @@ plan.addRule(new BackupPlanRule({
+## Importing existing backup vault
+
+To import an existing backup vault into your CDK application, use the `BackupVault.fromBackupVaultArn` or `BackupVault.fromBackupVaultName` 
+static method. Here is an example of giving a Lambda function permission to start a backup job:
```
```suggestion
static method. Here is an example of giving a IAM role permission to start a backup job:
```

BenChaimberg commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -108,27 +121,75 @@ export enum BackupVaultEvents {
+      throw new Error(`Backup Vault Arn ${backupVaultArn} does not have a resource name.`);
+    }
+
+    const backupVaultName = parsedArn.resourceName;
+
+    class Import extends BackupVaultBase {
       public readonly backupVaultName = backupVaultName;
```
```suggestion
    class Import extends BackupVaultBase {
      public readonly backupVaultName = parsedArn.resourceName;
```

BenChaimberg reviewed (changes requested):
Please also add more detail to the PR body about what it contains as it will be recorded in the squashed commit

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 21e9c0487d7cf436d1f1559eba8699d451e7274d
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3bfcf564-a2ee-444e-ac80-9f824df9275d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

DaWyz commented:
@BenChaimberg thanks for the review. I updated the code. I'm still a bit confused with the fixture. I have updated it and created a new one with a plan.

The readme was not compiling before. that's why I added all the `declare const ...` in the fixture.
In order to avoid that, I need to create the other constructs (e.g. topic, key, etc..) in the `README.md`. Let me know if it's fine.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 3cfe58cf433d1da8b66effdf23f4dae87ad6038b
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=18e99ad7-e58d-45de-bf08-1409b23f305e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: c9bc4591a5211e22235deee41718ded1b93a885d
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2574d904-51d3-45db-8357-6f7d9d3b5e22%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: c02188014cad8dd5b9f3c8d91a997d84fe509ae4
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a3332f47-e50f-4e14-96b9-6e8b4a30bc17%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
